test/multiplex_test.rb

```diff
@@ -0,0 +1,51 @@
+require_relative 'test_helper'
+
+class GraphQL::MultiplexTest < Minitest::Test
+  attr_reader :queries
+
+  def setup
+    @queries = []
+    QueryNotifier.subscriber = ->(query) { @queries << query }
+  end
+
+  def teardown
+    QueryNotifier.subscriber = nil
+  end
+
+  def schema_multiplex(*args)
+    ::Schema.multiplex(*args)
+  end
+
+
+  def test_batched_find_by_id
+    query_string = <<-GRAPHQL
+      query FetchTwoProducts($id1: ID!, $id2: ID!){
+        first: product(id: $id1) { id, title }
+        second: product(id: $id2) { id, title }
+      }
+    GRAPHQL
+
+    results = schema_multiplex([
+      { query: query_string, variables: { "id1" => "1", "id2" => "2"} },
+      { query: query_string, variables: { "id1" => "1", "id2" => "3"}},
+    ])
+
+    expected_1 = {
+      "data" => {
+        "first" => { "id" => "1", "title" => "Shirt" },
+        "second" => { "id" => "2", "title" => "Pants" },
+      }
+    }
+
+    expected_2 = {
+      "data" => {
+        "first" => { "id" => "1", "title" => "Shirt" },
+        "second" => { "id" => "3", "title" => "Sweater" },
+      }
+    }
+
+    assert_equal expected_1, results.first
+    assert_equal expected_2, results.last
+    assert_equal ["Product/1,2,3"], queries
+  end
```

Comment:
That's a great point, and it's something I overlooked. It makes no special treatment for mutations, but it could. (Apparently the JS server doesn't make special treatment either: https://github.com/apollographql/graphql-server/issues/176#issuecomment-290956816)

Maybe we could do a flow like this: 

- Shift the next operation from the set
  - If it's a mutation:
    - Run any already-gathered queries concurrently
    - Run it in isolation 
  - If it's a query, add it to the next multiplex
- When the set is empty: 
  - Run any already-gathered queries concurrently 

I think that would maintain the expected ordering for mutations: queries before mutations would not reflect their changes, but queries after mutations _would_ reflect mutations. 

However, now that I write it, I don't know if the incoming queries _are_ ordered. Maybe they come as an unordered set. 

If that's the case, maybe a better flow is mutate-then-read: 

- For each operation in the set, if it's a mutation, run its root fields
- Then, fullfill all other fields concurrently. 

That way, all reads in the multiplex reflect the changes of all mutations in the multiplex. 

Hmm, what if the mutations are order-dependent? :head_explodes: